dbms/src/Interpreters/Join.cpp

```diff
@@ -2025,6 +2114,28 @@ void Join::joinTotals(Block & block) const
     }
 }
 
+void Join::setProbeConcurrencyAndMaxBlockSize(size_t probe_concurrency_, UInt64 max_block_size)
+{
+    if (unlikely(probe_concurrency || max_block_size_for_hash_join))
+        throw Exception("Logical error: `setProbeConcurrencyAndMaxBlockSize` shouldn't be called more than once", ErrorCodes::LOGICAL_ERROR);
+    probe_concurrency = std::max(1, probe_concurrency_);
+    for (size_t i = 0; i < probe_concurrency; i++)
+    {
+        probe_process_infos.push_back(std::make_shared<ProbeProcessInfo>(Block{}, max_block_size, 0, 0, true, false));
+    }
+    max_block_size_for_hash_join = max_block_size;
+    probe_initialized = true;
+}
+
+void Join::setBlockAndInitProbeProcessInfo(Block block, size_t probe_index)
+{
+    probe_process_infos[probe_index]->block = block;
+    probe_process_infos[probe_index]->start_row = 0;
+    probe_process_infos[probe_index]->end_row = 0;
+    probe_process_infos[probe_index]->all_rows_joined_finish = true;
+    probe_process_infos[probe_index]->block_full = false;
+}
```

Comment:
move to ProbeProcessInfo?
```

ProbeProcessInfo::setAndInit(Block && block_)
{
    block = std::move(block_);
    start_row = 0;
    end_row = 0;
    all_rows_joined_finish = true;
    block_full = false;
}
```

Reply:
And pass block by right-value references `Block &&`

Reply:
done.